klarna-on-demand/src/main/java/com/klarna/ondemand/CryptoImpl.java

```diff
@@ -0,0 +1,116 @@
+import java.security.Key;
+import java.security.KeyFactory;
+import java.security.KeyPair;
+import java.security.KeyPairGenerator;
+import java.security.NoSuchAlgorithmException;
+import java.security.PrivateKey;
+import java.security.PublicKey;
+import java.security.Signature;
+import java.security.SignatureException;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.security.spec.X509EncodedKeySpec;
+
+class CryptoImpl implements Crypto {
+    private static final String PUBLIC_KEY = "PublicKey";
+    private static final String PRIVATE_KEY = "PrivateKey";
+    private static final String PREFERENCES_FILE_NAME = "KeyPair";
+    private static final String ALGORITHM = "RSA";
+
+    private static CryptoImpl cyptoInstance;
+
+    private String publicKeyBase64Str;
+    private PublicKey publicKey;
+    private PrivateKey privateKey;
+
+
+    public static Crypto getInstance(android.content.Context context) {
+        if (cyptoInstance == null) {
+            try {
+                cyptoInstance = new CryptoImpl(context);
+            } catch (Exception e) {
+                throw new RuntimeException("Could not initialize " + CryptoImpl.class.getName(), e);
+            }
+        }
+        return cyptoInstance;
+    }
+
+    private CryptoImpl(android.content.Context context) throws NoSuchAlgorithmException, InvalidKeySpecException {
+        SharedPreferences sharedPreferences = context.getSharedPreferences(PREFERENCES_FILE_NAME, Context.MODE_PRIVATE);
+
+        publicKey = readPublicKey(sharedPreferences);
+        privateKey = readPrivateKey(sharedPreferences);
+        if (publicKey == null || privateKey == null) {
+            KeyPair keyPair = generateKeyPair();
+            publicKey = keyPair.getPublic();
+            privateKey = keyPair.getPrivate();
+
+            persistKeyPair(sharedPreferences, keyPair);
+        }
+
+        publicKeyBase64Str = new String(Base64.encode(publicKey.getEncoded(), Base64.DEFAULT));
+    }
+
+    @Override
+    public String getPublicKeyBase64Str() {
+        return publicKeyBase64Str;
+    }
+
+    @Override
+    public String sign(String message) throws NoSuchAlgorithmException, InvalidKeyException, SignatureException {
+        Signature sign = Signature.getInstance("SHA256withRSA");
+        sign.initSign(privateKey);
+        sign.update(message.getBytes());
+        return new String(Base64.encode(sign.sign(), Base64.DEFAULT));
+    }
+
+    private void persistKeyPair(SharedPreferences sharedPreferences, KeyPair keyPair) {
+        SharedPreferences.Editor sharedPreferencesEditor = sharedPreferences.edit();
+
+        persistKey(sharedPreferencesEditor, PUBLIC_KEY, keyPair.getPublic());
+        persistKey(sharedPreferencesEditor, PRIVATE_KEY, keyPair.getPrivate());
+
+        sharedPreferencesEditor.commit();
+    }
+
+    private void persistKey(SharedPreferences.Editor sharedPreferencesEditor, String keyName, Key key) {
```

Comment:
As we discussed, there's a problem with `persistKey`, because it requires `sharedPreferencesEditor.commit();` to be called afterwards.
1. Extract 'byte[] keyBytes = key.getEncoded();  String keyStr = new String(Base64.encode(keyBytes, Base64.DEFAULT));'  to a different method.
2. remove `persistKey`.